File: tokutils.py
"""
Utility module for Cardano Token manipulation
"""
from os import makedirs, path
from subprocess import run as subprocess_run 
from collections import defaultdict
from json import dump as json_dump, loads as json_loads

def calculate_tokens_balance(tokens):
  """
  compute total balance for each list of tokens (token,amount)
  return a dict with total amount for each token
  """
  result = defaultdict(int)
  for key, value in tokens:
    result[key] += value
  return result

def create_keypair(address_type, addresses_path, address_prefix, name):
  """
  create keypair based on address_name
  """
  vkey_file = get_vkey_file(addresses_path, address_prefix, name)
  skey_file = get_skey_file(addresses_path, address_prefix, name)

  if(path.exists(vkey_file)) :
    print(address_prefix, "key pair already exists for", name)
    return
  
  makedirs(path.dirname(vkey_file), mode=0o777, exist_ok=True)

  run_params = ['cardano-cli', address_type, 'key-gen', '--verification-key-file', vkey_file, '--signing-key-file', skey_file]
  subprocess_run(run_params, capture_output=False, text=True)
  return

def create_policy(policy_name, policy_path):
  """
  create policy
  """
  policy_script=policy_path+'policy.script'
  policy_vkey=policy_path+'policy.vkey'
  policy_skey=policy_path+'policy.skey'
  policy = {}
  policy['policy_script'] = policy_script
  policy['policy_vkey'] = policy_vkey
  policy['policy_skey'] = policy_skey

  # check if token exists
  # if so, returns existing policy
  if path.exists(policy_script) :
    print("Policy exists : no policy created for", policy_name)
    policy_id = subprocess_run(['cardano-cli', 'transaction', 'policyid', '--script-file', policy_script], capture_output=True)
    policy['policy_id'] = policy_id.stdout.decode().replace('\n', '')
    return policy

  makedirs(policies_folder+policy_name, mode=0o777, exist_ok=True)

  rc = subprocess_run(['cardano-cli', 'address', 'key-gen', '--verification-key-file', policy_vkey, '--signing-key-file', policy_skey], capture_output=False)

  # create policy script
  keyhash = subprocess_run(['cardano-cli', 'address', 'key-hash', '--payment-verification-key-file', policy_vkey], capture_output=True, text=True)
  data = {}
  data['keyHash'] = keyhash.stdout.replace('\n', '')
  data['type'] = 'sig'
  with open(policy_script, 'w') as outfile:
    json_dump(data, outfile)

  # get policy id
  policy_id = subprocess_run(['cardano-cli', 'transaction', 'policyid', '--script-file', policy_script], capture_output=True, text=True)

  policy['policy_id'] = policy_id.stdout.replace('\n', '')

  return policy

def get_address(address_file):
  """
  get address from file
  """
  if not path.exists(address_file) :
    print("file not found :", address_file)
    return None
  addr_file = open(address_file,'r')
  address = addr_file.readlines()
  return address[0]

def get_policy(policy_name, policy_path):
  """
  get policy
  """
  if policy_name is None:
    return {}
  policy_script=policy_path+'policy.script'
  policy_vkey=policy_path+'policy.vkey'
  policy_skey=policy_path+'policy.skey'

  policy = {}
  policy['policy_script'] = policy_script
  policy['policy_vkey'] = policy_vkey
  policy['policy_skey'] = policy_skey

  # check if policy script exists
  # if so, returns existing policy
  if(path.exists(policy_script)) :
    run_params = ['cardano-cli', 'transaction', 'policyid', '--script-file', policy_script]
    policy_id = subprocess_run(run_params, capture_output=True)
    policy['policy_id'] = policy_id.stdout.decode().replace('\n', '')
    return policy
  else:
    return {}

def get_policy_id(token_name, utxo):
  """
  retrieve policy id from token name in utxo
  """
  assets_id = [k.split('.') for k in utxo['balances'].keys() if len(k.split('.')) == 2 and k.split('.')[1] == token_name]
  if len(assets_id) == 1:
    policy_id = assets_id[0][0]
  else:
    policy_id = None
  return policy_id

def get_protocol_keydeposit(network):
  """
  get keyDeposit parameter from protocol
  """
  network_name = network['network']
  network_magic = str(network['network_magic'])
  network_era = network['network_era']
  env_param = network['env']
  rc = subprocess_run(['cardano-cli', 'query', 'protocol-parameters', network_name, network_magic, network_era], \
    capture_output=True, text=True, env=env_param)
  return int(json_loads(rc.stdout)['keyDeposit'])

def get_protocol_parameters(network, protparams_file):
  """
  get protocol parameters
  """
  network_name = network['network']
  network_magic = str(network['network_magic'])
  network_era = network['network_era']
  env_param = network['env']
  subprocess_run(['cardano-cli', 'query', 'protocol-parameters', network_name, network_magic, network_era, '--out-file', protparams_file], \
    capture_output=False, text=True, env=env_param)
  return 

def get_policy_path(address_path, owner, policy_name, policies_folder):
  return get_address_path(address_path, owner)+policies_folder+policy_name+'/'

def get_address_path(addresses_path, name):
  return addresses_path+name+'/'

def get_address_key_file(addresses_path, address_type, address_or_key, name):
  """
  give file name for name type address or key
  """
  if not address_type in ['payment', 'stake']:
    print('Unknown address type :', address_type)
    return None
  if address_or_key == 'address':
    ext = '.addr'
  elif address_or_key == 'signing_key':
    ext = '.skey'
  elif address_or_key == 'verification_key':
    ext = '.vkey'
  else:
    print('Unknown type :', address_or_key)
    return None

  addr_key_file = get_address_path(addresses_path, name)+address_type+name+ext
  return addr_key_file

def get_address_file(addresses_path, address_type, name):
  """
  give file name for name type address
  """
  return get_address_key_file(addresses_path, address_type, 'address', name)

def get_skey_file(addresses_path, address_type, name):
  """
  give file name for name type signing key
  """
  return get_address_key_file(addresses_path, address_type, 'signing_key', name)

def get_vkey_file(addresses_path, address_type, name):
  """
  give file name for name type verification key
  """
  return get_address_key_file(addresses_path, address_type, 'verification_key', name)

File: Tokutils.hs
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE DeriveGeneric #-}
module Tokutils ( createKeypair, createPolicy, Address, AddressType(Payment, Stake), BlockchainNetwork (BlockchainNetwork, network, networkMagic, networkEra, networkEnv), getPolicyPath, getPolicyId, getProtocolKeyDeposit, saveProtocolParameters, getAddress, getAddressFile, getSkeyFile, getVkeyFile ) where

import System.Directory ( createDirectoryIfMissing, doesFileExist)
import System.FilePath ( takeDirectory )
import System.IO ( hGetContents, readFile)
import System.Process ( createProcess, proc, std_out, StdStream(CreatePipe), waitForProcess )
import Data.Aeson (decode, encode)
import Data.Aeson.TH(deriveJSON, defaultOptions, Options(fieldLabelModifier))
import GHC.Generics
import qualified Data.ByteString.Lazy.Char8 as B

-- policy helpers ---------------------------------------------------------
data PolicyScript = PolicyScript
  { 
    keyHash :: String
  , keyType    :: String
  }
  deriving (Show, Generic)

$(deriveJSON defaultOptions{fieldLabelModifier = \f -> if f == "keyType" then "type" else f} ''PolicyScript)

-- Cardano Policy
data Policy = Policy
  { 
    policyScript :: FilePath
  , policyVkey   :: FilePath
  , policySkey   :: FilePath
  , policyId     :: String
  }
  deriving Show

-- build Policy full file names
buildPolicyPath :: String -> String -> Policy
buildPolicyPath policyName policyPath = Policy 
  (policyPath ++ "policy.script")
  (policyPath ++ "policy.vkey")
  (policyPath ++ "policy.skey")
  ""

-- create a Cardano Policy
createPolicy :: String -> String -> IO (Maybe Policy)
createPolicy policyName policyPath = do
  let policy = buildPolicyPath policyName policyPath
  -- check if policy script exists
  -- if so returns existing Policy
  bool <- doesFileExist (policyScript policy)
  if not bool then do
    createDirectoryIfMissing True policyPath
    -- create policy key files
    (_, Just rc, _, ph) <- createProcess (proc "cardano-cli" ["address", "key-gen", "--verification-key-file", policyVkey policy, "--signing-key-file", policySkey policy]){ std_out = CreatePipe }
    r <- waitForProcess ph
    -- create hash
    (_, Just hout, _, _) <- createProcess (proc "cardano-cli" ["address", "key-hash", "--payment-verification-key-file", policyVkey policy]){ std_out = CreatePipe }
    keyh <- hGetContents hout
    let keyhash = filter (/= '\n') keyh
    -- create policy script
    let myPolicyScript = PolicyScript { keyHash = keyhash, keyType = "sig"}
    B.writeFile (policyScript policy) (encode myPolicyScript)
  else do
    putStrLn $ "Policy exists : no policy created for " ++ policyName
  -- retrieve policy script
  (_, Just hout, _, _) <- createProcess (proc "cardano-cli" ["transaction", "policyid", "--script-file", policyScript policy]){ std_out = CreatePipe }
  pId <- hGetContents hout 
  return (Just policy { policyId = filter (/= '\n') pId} )

-- get Policy Folder
getPolicyPath:: FilePath -> String -> String -> FilePath -> FilePath
getPolicyPath addressPath ownerName policyName policiesFolder = getAddressPath addressPath ownerName ++ policiesFolder ++ policyName ++ "/"

-- get Policy Id
getPolicyId:: Policy -> String
getPolicyId = policyId

-- protocols helpers ------------------------------------------------------

data BlockchainNetwork = BlockchainNetwork 
  {
    network :: String
  , networkMagic :: Int
  , networkEra :: String
  , networkEnv :: String
  }
  deriving Show

-- get keyDeposit parameter from protocol
getProtocolKeyDeposit :: BlockchainNetwork -> IO (Maybe Int)
getProtocolKeyDeposit bNetwork = do
  let netName = network bNetwork
  let netMagic = networkMagic bNetwork
  let netEra = networkEra bNetwork
  let envParam = networkEnv bNetwork
  (_, Just rc, _, ph) <- createProcess (proc "cardano-cli" ["query", "protocol-parameters", netName, show netMagic, netEra]){ std_out = CreatePipe }
  r <- waitForProcess ph
  jsonFile <- hGetContents rc
  js <- B.readFile jsonFile
  let jsa = decode js
  print jsa
  -- TODO Retrieve keyDeposit from file
  -- return int(json_loads(rc.stdout)['keyDeposit'])
  return jsa

-- get protocol parameters
saveProtocolParameters :: BlockchainNetwork -> FilePath -> IO Bool
saveProtocolParameters bNetwork protocolParams = do
  let netName = network bNetwork
  let netMagic = networkMagic bNetwork
  let netEra = networkEra bNetwork
  let envParam = networkEnv bNetwork
  (_, Just rc, _, ph) <- createProcess (proc "cardano-cli" ["query", "protocol-parameters", netName, show netMagic, netEra, "--out-file", protocolParams]){ std_out = CreatePipe }
  r <- waitForProcess ph
  return True

-- create keypair based on address_name
createKeypair :: AddressType -> FilePath -> String -> IO Bool
createKeypair addressType addressesPath ownerName = do
  let vkeyFile = getVkeyFile addressesPath addressType ownerName
  let skeyFile = getSkeyFile addressesPath addressType ownerName
  bool <- doesFileExist vkeyFile
  if bool then do
    putStrLn $ "key pair already exists for " ++ ownerName
    return False
  else do
    createDirectoryIfMissing True (takeDirectory vkeyFile)
    let saddressType = if addressType == Payment then "address" else "stake-address"
    
    (_, Just rc, _, ph) <- createProcess (proc "cardano-cli" [saddressType, "key-gen", "--verification-key-file", vkeyFile, "--signing-key-file", skeyFile]){ std_out = CreatePipe }
    r <- waitForProcess ph
    return True

-- calculate_tokens_balance

-- address helpers ---------------------------------------------------------------------

-- data AdressOrKey = address | signing_key | verification_key deriving (Read, Show, Eq)
type Address = String
data AddressType = Payment | Stake deriving (Read, Show, Eq)

-- TODO GetAddress
getAddress :: FilePath -> IO (Maybe Address)
getAddress addressFileName = do
  bool <- doesFileExist addressFileName
  if not bool then do
    putStrLn $ "file not found : " ++ addressFileName
    return Nothing
  else do 
    addr <- readFile addressFileName
    return (Just addr)

-- compute address path from addresses path and owner name
getAddressPath:: FilePath -> String -> FilePath
getAddressPath addressesPath ownerName = addressesPath ++ ownerName ++ "/"

-- give file name for name type address or key
getAddressKeyFile :: FilePath -> AddressType -> String -> String -> FilePath
getAddressKeyFile addressesPath addressType addressKey name = do
  let saddressType = if addressType == Payment then "payment" else "stake"
  let extmap = [ ("address", ".addr"), ("signing_key", ".skey"), ("verification_key", ".vkey")]
  let extm = lookup addressKey extmap
  case extm of
    Just ext -> getAddressPath addressesPath name ++ saddressType ++ ext
    _ -> ""

-- give file name for name type address
getAddressFile :: FilePath -> AddressType -> String -> FilePath
getAddressFile addressesPath addressType = getAddressKeyFile addressesPath addressType "address"

-- give file name for name type signing key
getSkeyFile :: FilePath -> AddressType -> String -> FilePath
getSkeyFile addressesPath addressType = getAddressKeyFile addressesPath addressType "signing_key"

-- give file name for name type verification key
getVkeyFile :: FilePath -> AddressType -> String -> FilePath
getVkeyFile addressesPath addressType = getAddressKeyFile addressesPath addressType "verification_key"
